src/gt4py/next/iterator/type_system/type_synthesizer.py

```diff
@@ -280,6 +280,20 @@ def applied_as_fieldop(*fields) -> ts.FieldType:
     return applied_as_fieldop
 
 
+@_register_builtin_type_synthesizer
+def cond(
+    pred: ts.ScalarType, true_branch: ts.DataType, false_branch: ts.DataType
+) -> TypeSynthesizer:
+    assert isinstance(pred, ts.ScalarType) and pred.kind == ts.ScalarKind.BOOL
+    assert true_branch == false_branch
+
+    @TypeSynthesizer
+    def apply_cond() -> ts.DataType:
+        return true_branch
+
+    return apply_cond
```

Comment:
```suggestion
@_register_builtin_type_synthesizer
def cond(
    pred: ts.ScalarType, true_branch: ts.DataType, false_branch: ts.DataType
) -> TypeSynthesizer:
    assert isinstance(pred, ts.ScalarType) and pred.kind == ts.ScalarKind.BOOL
    assert true_branch == false_branch
    assert isinstance(true_branch, (ts.FieldType, ts.DeferredType))
    
    return true_branch
```
I'm a little confused why the `cond` signature is `cond(cond_, true_branch, false_branch)()` and not `cond(cond_, true_branch, false_branch)`. What is this extra call for? Maybe something to discuss on zoom.